### PR TITLE
Check fflush on BIO_ctrl call

### LINKED ISSUE #3232: Writing to a full disk does not result in "no space left on device"
Writing to a file on a full disk should produce an error message or at least a non-zero exit code. However, with version 1.0.2k (and equally with the latest version from master 7db0289ea70b9bec94e32350cc8ec9da96ab2776) the following happens:
```
% df /tmp/small                                                           
Filesystem     1K-blocks  Used Available Use% Mounted on
none                   4     4         0 100% /tmp/small
% openssl aes-256-cbc -a -pass pass:pass -in /dev/null                    
U2FsdGVkX1/uWnU+dbf+vINloOWFI6NpaZ3YfKCjJCg=
% openssl aes-256-cbc -a -pass pass:pass -in /dev/null -out /tmp/small/out
% echo $?
0
% cat /tmp/small/out                                                      
% openssl aes-256-cbc -a -pass pass:pass -in /dev/null | cat > /tmp/small/out
cat: write error: No space left on device
```

### PR DESCRIPTION
**Pull request by richsalz**
Bug found and fix suggested by Julian Rüth.

Fixes #3232

### CODE REVIEW

File: crypto/bio/bss_file.c [hunk 1]
```diff
@@ -319,8 +319,8 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
     case BIO_CTRL_FLUSH:
         if (b->flags & BIO_FLAGS_UPLINK)
             UP_fflush(b->ptr);
```
Comment (t-j-h) [inline crypto/bio/bss_file.c]: If you are going to check the fflush return information it should also be done here in the uplink case too

Comment (richsalz) [inline crypto/bio/bss_file.c]: Does that act just like "regular" fflush?  Okay, updated commit coming shortly.

**Comment by richsalz** [comment]
updated commit pushed

**Comment by kroeckx** [comment]
Should we do something with errno? Add it to the error queue?

File: crypto/bio/bss_file.c [hunk 2]
```diff
@@ -317,10 +317,11 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
...
     case BIO_CTRL_FLUSH:
-        if (b->flags & BIO_FLAGS_UPLINK)
-            UP_fflush(b->ptr);
-        else
-            fflush((FILE *)b->ptr);
+        if (b->flags & BIO_FLAGS_UPLINK) {
+            if (UP_fflush(b->ptr) == EOF)
+                ret = 0;
+        } else if (fflush((FILE *)b->ptr) == EOF)
+            ret = 0;
```
Comment (levitte) [inline crypto/bio/bss_file.c]: { and } around this body as well, please

**Comment by richsalz** [comment]
updated commit pushed.

**Comment by richsalz** [comment]
@kroeckx was your errno comment a question or a request to change this PR?

**Comment by kroeckx** [comment]
I'm wondering what we do in other places, and we at least things like:
SYSerr(SYS_F_FOPEN, get_last_sys_error());


Kurt

**Comment by richsalz** [comment]
I just pushed another commit that does this.  I think instead we should have a ```SYS_F_STDIO``` and use that instead.  We don't have ```SY'S_F_xxx``` for most stdio functions where we'd like to push an error code.  What do you think?

**Comment by levitte** [comment]
I think you should add the necessary stuff for SYS_F_FFLUSH. Unless I remember wrong, it's three places to change, one added line each.

**Comment by levitte** [comment]
Ah, you did already...

File: crypto/bio/bss_file.c [hunk 3]
```diff
@@ -317,10 +318,14 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
...
     case BIO_CTRL_FLUSH:
-        if (b->flags & BIO_FLAGS_UPLINK)
-            UP_fflush(b->ptr);
-        else
-            fflush((FILE *)b->ptr);
+        st = b->flags & BIO_FLAGS_UPLINK
+                ? UP_fflush(b->ptr) : fflush((FILE *)b->ptr);
+        if (st == EOF) {
+            SYSerr(SYS_F_FFLUSH, get_last_sys_error());
+            ERR_add_error_data(5, "fopen('", ptr, "','", p, "')");
```
Comment (levitte) [inline crypto/bio/bss_file.c]: This line seems wrong...

Comment (richsalz) [inline crypto/bio/bss_file.c]: fixed; updated commit pushed.

**Comment by richsalz** [comment]
ping @levitte and @kroeckx

File: crypto/bio/bss_file.c [hunk 4]
```diff
@@ -317,10 +318,14 @@ static long file_ctrl(BIO *b, int cmd, long num, void *ptr)
...
     case BIO_CTRL_FLUSH:
-        if (b->flags & BIO_FLAGS_UPLINK)
-            UP_fflush(b->ptr);
-        else
-            fflush((FILE *)b->ptr);
+        st = b->flags & BIO_FLAGS_UPLINK
+                ? UP_fflush(b->ptr) : fflush((FILE *)b->ptr);
+        if (st == EOF) {
+            SYSerr(SYS_F_FFLUSH, get_last_sys_error());
+            ERR_add_error_data(3, "fflush('", ptr, "')");
```
Comment (levitte) [inline crypto/bio/bss_file.c]: All args but the first must be C strings. `ptr` isn't...

Comment (richsalz) [inline crypto/bio/bss_file.c]: fixed, updated commit pushed.

**Comment by richsalz** [comment]
PR #3348 for 1.0.2, 595b2a4 in master and 913d3a6 in 1.1.0

**Pull request closed without merging**